fix offline build mode

lethliel opened the pull request:
The offline mode does not really work offline since
parse_repoarchdescr() uses get_buildconfig to store the
buildconfig in a temporary file.

Solution:

Use the same logic as in build.py. If in offline mode(noinit = True)
try to use the local _buildconfig file. If not in offline mode just
download the buildconfig from the server via get_buildconfig

fixes https://github.com/openSUSE/osc/issues/560

marcus-h commented:
On 2019-05-21 05:35:02 -0700, Marco Strigl wrote:
> diff --git a/osc/commandline.py b/osc/commandline.py
> index 05faaac1..b24d9cdb 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -6092,16 +6092,29 @@ def parse_repoarchdescr(self, args, noinit = False, alternative_project = None,
>                      raise oscerr.WrongArgs("Repository is missing. Cannot guess build description without repository")
>                  apiurl = self.get_api_url()
>                  project = store_read_project('.')
> -                bc = get_buildconfig(apiurl, project, arg_repository)
> -                with tempfile.NamedTemporaryFile() as f:
> -                    f.write(bc)
> -                    f.flush()
> -                    # some distros like Debian rename and move build to obs-build
> -                    if not os.path.isfile('/usr/lib/build/queryconfig') and os.path.isfile('/usr/lib/obs-build/queryconfig'):
> -                        queryconfig = '/usr/lib/obs-build/queryconfig'
> +                if not os.path.isfile('/usr/lib/build/queryconfig') and os.path.isfile('/usr/lib/obs-build/queryconfig'):
> +                    queryconfig = '/usr/lib/obs-build/queryconfig'
> +                else:
> +                    queryconfig = '/usr/lib/build/queryconfig'
> +                if noinit:
> +                    bc_filename = '_buildconfig-%s-%s' % (arg_repository, arg_arch)
> +                    if is_package_dir('.') and os.access(osc.core.store, os.W_OK):

Hmm why do you check for write permission?

> +                        bc_filename = os.path.join(os.getcwd(), osc.core.store, bc_filename)
> +                    elif not os.access('.', os.W_OK):
> +                        bc_file = NamedTemporaryFile(prefix=bc_filename)
> +                        bc_filename = bc_file.name

Hmm why? We are not going to write to this file. Actually, there
is no need to create a tmpfile.

>                      else:
> -                        queryconfig = '/usr/lib/build/queryconfig'
> -                    recipe = return_external(queryconfig, '--dist', f.name, 'type')
> +                        bc_filename = os.path.abspath(bc_filename)
> +                    if not os.path.isfile(bc_filename):
> +                        raise oscerr.WrongOptions('--offline is not possible, no local buildconfig file')
> +                    recipe = return_external(queryconfig, '--dist', bc_filename, 'type')
> +                else:
> +                    bc = get_buildconfig(apiurl, project, arg_repository)
> +                    with tempfile.NamedTemporaryFile() as f:
> +                        f.write(bc)
> +                        f.flush()
> +                        # some distros like Debian rename and move build to obs-build

Hmm the comment belongs to the "queryconfig =..." lines above.

> +                        recipe = return_external(queryconfig, '--dist', f.name, 'type')
>                  recipe = recipe.strip()
>                  if recipe == 'arch':
>                      recipe = 'PKGBUILD'

lethliel commented:
@marcus-h 

* moved the comment to the right place
* removed unneeded parts in the bc_filename creation (In the first commit I just copied the complete part from build.py)
* note to myself: Check after copy & paste ;-)

marcus-h commented:
On 2019-05-22 05:00:53 -0700, Marco Strigl wrote:
> * note to myself: Check after copy & paste ;-)
>
:)

Merged. Thanks!
